### PR TITLE
use trainable activation quantizers in gptq (pytorch)

### PR DESCRIPTION
**Pull request by irenaby**
## Pull Request Description:
GPTQ used inferable activation quantizers which in pytorch do not propagate gradients, thus not all gradients reached the weights. 
Added new trainable activation quantizers without torch.no_grad for POT, symmetric and uniform. 

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: tests/pytorch_tests/function_tests/test_gptq_activation_quantizers.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        self._run_test(q_inferable, q, [QuantizationMethod.POWER_OF_TWO, QuantizationMethod.SYMMETRIC])
+
+    def test_uniform(self):
+        kwargs = dict(num_bits=5, min_range=[-3], max_range=[1.5])
+        q_inferable = ActivationUniformInferableQuantizer(**kwargs)
+        q = STEActivationUniformGPTQTrainableQuantizer(**kwargs)
+        self._run_test(q_inferable, q, [QuantizationMethod.UNIFORM])
+
+    def _run_test(self, q_infer, q, exp_q_method):
+        x = torch.randn((5, 10, 20), requires_grad=True, generator=torch.Generator().manual_seed(42))
```
Comment (ofirgo) [inline tests/pytorch_tests/function_tests/test_gptq_activation_quantizers.py:40]: what does the generator=torch.Generator() do? (just want to know cause I'm not familiar with this)

File: tests/pytorch_tests/function_tests/test_gptq_activation_quantizers.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        y_infer = q_infer(x)
+        y = q(x)
+        self.assertTrue(torch.equal(y_infer, y))
+        # if autograd cannot propagate, will crash
```
Comment (ofirgo) [inline tests/pytorch_tests/function_tests/test_gptq_activation_quantizers.py:44]: will it crash without an assertion? if it will then wrap it with assertRaise and check that it is crashed because of the correct exception

**Review by ofirgo — CHANGES_REQUESTED** [review]
see some comments and let's talk offline if there are any questions

File: model_compression_toolkit/gptq/common/gptq_config.py [hunk 3]
```diff
@@ -32,6 +32,10 @@ class RoundingType(Enum):
     SoftQuantizer = 1
 
 
+class ActivationGradProp(Enum):
```
[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/gptq/common/gptq_config.py:35]: I considered that but then there will be two different quantizers (qat and the new one) that will meet the query during quantizer discovery. I agree that defining identifiers that are supposed to be unique locally in multiple places is not a good approach. Quantizers discovery can definitely be a topic on our sw discussions list.

File: model_compression_toolkit/gptq/pytorch/quantizer/activation/ste_activation.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+@author: irenab
+"""
+import torch
+from mct_quantizers import mark_quantizer, QuantizationTarget, QuantizationMethod
+from mct_quantizers.pytorch.quantizers import ActivationSymmetricInferableQuantizer, ActivationUniformInferableQuantizer
+
+from model_compression_toolkit.gptq.common.gptq_config import ActivationGradProp
+
+
+class BaseActivationPytorchGPTQTrainableQuantizer:
```
[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/gptq/pytorch/quantizer/activation/ste_activation.py:27]: Yes, there is a reason (as I wrote in the docstring). The more I think about it the more I'm convinced we are taking the wrong approach here. Currently there is a clear separation between inferable and trainable quantizers in mct: trainable learn their parameters and as such have a specific training-relevant interface. Inferable do not, and do not have such interface. I think this makes perfect sense. 

The new quantizer breaks this functional distinction completely - it wants to be trainable when there is really nothing trainable about it. It's not only about having to implement irrelevant method (even if it's just noop). It's also expected to have quantization config which is validated by BaseTrainableQuantizer (which again I think is a good thing). To make this work we need either to relax BaseTrainable to allow it to actually be non-learnable, or fake learnable api in the new quantizer (which unlike noop methods is not straightforward in itself). Both options are pretty horrible IMO.
If we want to have inferable allowed in training separated from just inferable, I believe my solution is the least nasty between the two. It defines a new type of quantizer that doesn't have to fake being Trainable, and extends the inferable quantizer for the implementation.

However, I still think we might be trying to solve the wrong problem. There is no inherent coupling between whether a layer can run in training and whether it learns its params. All the mess arises from the fact that we forcefully equate the two, and which mct is not designed for (and for good reasons). I think we should keep the definition of trainable as learnable, and inferable as not having learnable params. The rest is up to the user as in any other layer. 

* Practically, I don't see how torch.nograd helps the user. If inference is run properly with nograd() or none of the quantizer's inputs have require_grad, there won't be any autograd overhead whether we use nograd in quantizer or not. However if the user intentionally runs training, best case we cause it to crash, worst case is an unexpected behavior unbeknown to the user. Preventing overhead in a scenario that a user runs inference without disabling grad seems like a weird requirement, especially at the cost of a potentially unexpected behavior.

* Conceptually, if a user really wants to continue fine tuning a model after mct, why shouldn't we allow it? How is it different from BatchNorm that is sometimes being disabled for learning during fine-tuning while still propagating gradients, or any other weights with requires_grad False for that matter? I think once the model leaves mct it's just another layer and the user should have control over it.

Please correct me if I got anything wrong.

[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/gptq/pytorch/quantizer/activation/ste_activation.py:27]: And BTW it is not even consistent within the inferable quantizers. Quantization wrapper freezes the weights as it should, but I don't see anything that prevents grads from back propagation. And my understanding is that in keras there is no such thing at all.

File: tests/pytorch_tests/function_tests/test_gptq_activation_quantizers.py (hunk 1 above)
Comment (irenaby) [reply tests/pytorch_tests/function_tests/test_gptq_activation_quantizers.py:40]: Creates a standalone random generator. It's state is not not affected by any other random calls
